### PR TITLE
Interaction with portable lamps

### PR DESCRIPTION
**Pull request by SzwedzikPL**
demo: https://www.youtube.com/watch?v=VsEnMXXO-pY :D

### CODE REVIEW

**Comment by jonpas** [comment]
Would this be better in it's own module?

**Comment by SzwedzikPL** [comment]
I think it fits in interaction, basically is only one option.

**Comment by Glowbal** [comment]
I do like this.

File: addons/interaction/CfgVehicles.hpp [hunk 1]
```diff
@@ -604,4 +604,97 @@ class CfgVehicles {
...
+                condition = "true";
+                class ACE_SwitchLamp {
+                    displayName = CSTRING(TurnOn);
+                    condition = QUOTE(alive _target);
+                    statement = QUOTE([ARR_1(_target)] call DFUNC(switchLamp));
+                    selection = "";
+                    distance = 2;
+                    showDisabled = 0;
+                    priority = -1;
+                };
```
Comment (jonpas) [inline addons/interaction/CfgVehicles.hpp]: Use macros so you don't have to duplicate `ACE_Actions`?

Comment (Dobakjang) [inline addons/interaction/CfgVehicles.hpp]: NAVER - http://www.naver.com/
--------------------------------------------

kgino1045@naver.com 님께 보내신 메일 <Re: [ACE3] Interaction with portable lamps (#2230)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

Comment (SzwedzikPL) [inline addons/interaction/CfgVehicles.hpp]: Why question mark?

Comment (jonpas) [inline addons/interaction/CfgVehicles.hpp]: Because it's up to you, but it will make it cleaner.

**Comment by alganthe** [comment]
using a macro is easier than just typing the interaction code over and over.

File: addons/interaction/functions/fnc_switchLamp.sqf [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+private ["_vectors", "_pos", "_reflectors", "_hitpointsdmg"];
+
+_vectors = [vectorDir _object, vectorUp _object];
+_pos = getPosATL _object;
+
+_reflectors = (configfile >> "CfgVehicles" >> _objectClass >> "Reflectors") call BIS_fnc_getCfgSubClasses;
+_hitpointsdmg = [];
+{
+	_hitpoint = getText (configfile >> "CfgVehicles" >> _objectClass >> "Reflectors" >> _x >> "hitpoint");
```
Comment (Glowbal) [inline addons/interaction/functions/fnc_switchLamp.sqf]: Missing private

File: addons/interaction/functions/fnc_switchLamp.sqf [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+_class = getText (configFile >> "CfgVehicles" >> _objectClass >> QGVAR(switchLampClass));
+
+if (_class == "") exitWith {};
+
+private ["_vectors", "_pos", "_reflectors", "_hitpointsdmg"];
+
+_vectors = [vectorDir _object, vectorUp _object];
+_pos = getPosATL _object;
+
+_reflectors = (configfile >> "CfgVehicles" >> _objectClass >> "Reflectors") call BIS_fnc_getCfgSubClasses;
```
Comment (Glowbal) [inline addons/interaction/functions/fnc_switchLamp.sqf]:
```sqf
_reflectors = "true" configClasses (configfile >> "CfgVehicles" >> _objectClass >> "Reflectors");
```

**Comment by SzwedzikPL** [comment]
> using a macro is easier than just typing the interaction code over and over.

Yes, thank you :) I've added macro and other suggested changes.

File: addons/interaction/CfgVehicles.hpp [hunk 4]
```diff
@@ -1,25 +1,45 @@
...
+    class ACE_Actions { \
+        class ACE_MainActions { \
+            displayName = CSTRING(MainAction); \
+            selection = ""; \
+            distance = 2; \
+            condition = "true"; \
+            class ACE_SwitchLamp { \
+                displayName = CSTRING(var1); \
+                condition = QUOTE(alive _target); \
+                statement = QUOTE([ARR_1(_target)] call DFUNC(switchLamp)); \
```
Comment (jonpas) [inline addons/interaction/CfgVehicles.hpp]: `ARR_N` is unnecessary when you have only 1 element, `[_target] call ...` works.

Even better would be `_target call ...` without array.

File: addons/interaction/CfgVehicles.hpp [hunk 5]
```diff
@@ -1,25 +1,45 @@
...
-      };
-    };
-    class ModuleDescription {
-      description = CSTRING(Module_Description);
+    class ACE_Module;
+    class ACE_ModuleInteraction: ACE_Module {
+        author = ECSTRING(common,ACETeam);
+        category = "ACE";
+        displayName = CSTRING(Module_DisplayName);
+        function = "ACE_Interaction_fnc_moduleInteraction";
```
Comment (jonpas) [inline addons/interaction/CfgVehicles.hpp]: Seems like the best place to change, please change it to `QFUNC(moduleInteraction)`.

File: addons/interaction/functions/fnc_switchLamp.sqf [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+_newobject setPosATL _pos;
+_isOff = getNumber (configFile >> "CfgVehicles" >> _class >> QGVAR(switchLampOff)) == 1;
+
+if(_isOff) then {
+	//this lamp is off
+	{_newobject sethit [_x select 0, (_x select 1) max DISABLED_LAMP_DMG];nil} count _hitpointsdmg;
+} else {
+	//this lamp is on
+	{if((_x select 1) > DISABLED_LAMP_DMG) then {_newobject sethit _x;};nil} count _hitpointsdmg;
+};
```
Comment (jonpas) [inline addons/interaction/functions/fnc_switchLamp.sqf]: Maybe some comments in this function would be nice, most notably why is `GVAR(switchLampClass)` and `GVAR(switchLampOff)` needed and used for.

File: addons/interaction/CfgVehicles.hpp [hunk 7]
```diff
@@ -1,25 +1,45 @@
+#define MACRO_PORTABLELIGHTSACTIONS(var1) \
```
Comment (jonpas) [inline addons/interaction/CfgVehicles.hpp]: I'd use `MACRO_PORTABLE_LIGHTS_ACTION` for better readability.

**Comment by SzwedzikPL** [comment]
I've noticed that getHitpoints in common does not support reflector hitpoints and every time i drag or carry lamp reflector are fixing. @commy2 @Glowbal

File: addons/interaction/functions/fnc_switchLamp.sqf [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+private ["_vectors", "_pos", "_reflectors", "_hitpointsdmg"];
+
+_vectors = [vectorDir _object, vectorUp _object];
+_pos = getPosATL _object;
+
+_reflectors = "true" configClasses (configfile >> "CfgVehicles" >> _objectClass >> "Reflectors");
+_hitpointsdmg = [];
+{
+	private "_hitpoint";
```
Comment (jokoho48) [inline addons/interaction/functions/fnc_switchLamp.sqf]: TAAABs

**Comment by SzwedzikPL** [comment]
Lamps are working fine now but i found a problem with cargo. When im unloading a lamp from vehicle model is hidden:
http://i.imgur.com/sVR4qY6.jpg

`lamp setPosATL (getPosATL lamp)` make it visible again. I cant find a source of the problem.

File: addons/interaction/functions/fnc_switchLamp.sqf [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+private _reflectors = "true" configClasses (configfile >> "CfgVehicles" >> (typeof _lamp) >> "Reflectors");
+private _hitPointsDamage = [];
+{
+    private _hitPoint = getText (_x >> "hitpoint");
+    _hitPointsDamage pushback [_hitPoint, _lamp getHit _hitPoint];
+    nil
+} count _reflectors;
+
+//if lamp is on turn it off
+private _eventName = if(_isOn) then {"lampTurnOff"} else {"lampTurnOn"};
```
Comment (commy2) [inline addons/interaction/functions/fnc_switchLamp.sqf]: `private _eventName = ["lampTurnOn", "lampTurnOff"] select _isOn;`
for aesthetics

Comment (esteldunedain) [inline addons/interaction/functions/fnc_switchLamp.sqf]: > aesthetics

:picard:

Comment (commy2) [inline addons/interaction/functions/fnc_switchLamp.sqf]: Way better than `= if`

**Comment by SzwedzikPL** [comment]
I've removed some unnecessary code. Cargo bug still present but only if you're unloading lamp on ASL height below ~5.5

**Comment by Glowbal** [comment]
Thanks @SzwedzikPL

**Pull request merged**